### PR TITLE
Yiz/user logout

### PR DESCRIPTION
**Pull request by sleepyzzzzzz**
## Changes made in this PR
Please describe what change you've made in this branch.
implement CreateToken & InvalidateToken functions + specs in db store layer

## Corresponding noted Ruby branch
https://github.com/ParamountDataManagement/paramount-api/pull/1960

## TODO - changes that needs to be made
1. What to return from InvalidateToken?
2. return user access token record or just error? currently only error is return from InvalidateToken function
                                                                                        
## Implementation Checklist
Please check following check list to ensure the tasks are considered and completed in this branch.

#### API Design
- [ ] Defined gRPC API in proto file
- [ ] regenerated go code using `make proto_generate` command

#### API implementation
- [ ] Implemented API in corresponding gRPC server
- [ ] Added tests in gRPC layer

#### Domain Service and DB query Implementation
- [x] Added new query in dbstore layer
- [x] Added tests in db store layer
                                                                                       
- [ ] Implemented functionality in Domain Service Layer
- [ ] Added test in Domain Service Layer

- [x] Created a new branch in paramount-api, with comments there indicating the functionality in Ruby that's re-implemented in Golang

#### Postgres Layer - identify and potentially add db migration to create index for fast querying.
- [ ] Verified and potentially created index for the newly added query.

### CODE REVIEW

File: migrations/20220210055845_CreateUserAccessTokens.up.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
+create table if not exists user_access_tokens (
+  id serial primary key,
+  user_id integer not null,
+  changed_by_user_id integer,
```
Comment (lhan-the-real) [inline migrations/20220210055845_CreateUserAccessTokens.up.sql:4]: let's get rid of this field. It's not used at all.

File: migrations/20220210055845_CreateUserAccessTokens.up.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      foreign key (user_id) references users (id),
+  constraint FK_user_access_tokens_changed_by_users
+      foreign key (changed_by_user_id) references users (id)
+);
+
+create trigger update_user_access_tokens_expires_at 
+  before update on user_access_tokens 
+  for each row 
+  execute procedure updateExpireTimestamp();
+
```
Comment (lhan-the-real) [inline migrations/20220210055845_CreateUserAccessTokens.up.sql:19]: Is there an empty space in line 19? If so, let's remove the empty space in line19, but keep the newline.

File: migrations/20220210055845_CreateUserAccessTokens.up.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
+create table if not exists user_access_tokens (
```
Comment (lhan-the-real) [inline migrations/20220210055845_CreateUserAccessTokens.up.sql:1]: Let's add comments to each field of this table. Ignore the id, created_at, udpated_at field

File: migrations/20210128214728_CreateUpdatedTimestampFunction.up.sql [hunk 4]
```diff
@@ -6,4 +6,14 @@ $$
...
+language plpgsql;
+
+create or replace function updateExpireTimestamp()
+returns trigger as
+$$
+  begin
+    new.expires_at = (current_timestamp at time zone 'utc');
+    return new;
+  end;
+$$
```
Comment (lhan-the-real) [inline migrations/20210128214728_CreateUpdatedTimestampFunction.up.sql:18]: Add new line to the end of this file.

File: internal/domain/users/usersStore.go [hunk 5]
```diff
@@ -276,3 +281,64 @@ func isValidEmailAddress(email string) bool {
...
+	if rowsAffected == 0 {
+		return pdmerr.New(
+			"UsersStore: no row is updated for updating expires at timestamp",
+		).WithData(map[string]string{"token": accessToken, "user_agent_encrypted": userAgent})
+	}
+
+	return nil
+}
+
+func HexMD5HashString(token string) string {
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore.go:340]: move to utils.go. file

Comment (lhan-the-real) [reply internal/domain/users/usersStore.go:340]: Is it possible for us to add a test for this function?

File: internal/domain/users/domainService.go [hunk 6]
```diff
@@ -33,6 +33,16 @@ type UserInfo struct {
...
 }
 
+type UserAccessTokens struct {
+	ID                 int32         `db:"id"`
+	UserID             int32         `db:"user_id"`
+	ChangedByUserID    sql.NullInt32 `db:"changed_by_user_id"`
+	Token              string        `db:"token"`
+	ExpiresAt          time.Time     `db:"expires_at"`
+	UserAgent          string        `db:"user_agent"`
+	UserAgentEncrypted string        `db:"user_agent_encrypted"`
```
Comment (lhan-the-real) [inline internal/domain/users/domainService.go:43]: do we know where we use UserAgentEncrypted in the db?

File: internal/domain/users/usersStore.go (hunk 5 above)
Comment (lhan-the-real) [reply internal/domain/users/usersStore.go:340]: Let's also add a comment to describe the nature of MD5 hash. `same string encodes to the same hash, and it's not easily reversible `

File: internal/domain/users/usersStore_test.go [hunk 7]
```diff
@@ -214,6 +215,72 @@ var _ = Describe("UsersStore", func() {
...
+		Describe("Create Token", func() {
+			It("should successfully insert, and return the user access token as result", func() {
+				resultUserToken, err := usersStore.CreateToken(context.Background(), expectedUserInfo, userAgent)
+				Expect(err).To(BeNil())
+				Expect(resultUserToken.UserID).To(Equal(expectedUserInfo.ID))
+				Expect(resultUserToken.Token).NotTo(Equal(""))
+			})
+
+			It("should fail when inserting a user access token with an nonexistent user id", func() {
+				expectedUserInfo.ID = -1
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore_test.go:240]: If it's none existing, then let's don't use `-1`, let's use some really big number here.

File: internal/domain/users/domainService.go [hunk 8]
```diff
@@ -33,6 +33,16 @@ type UserInfo struct {
 	ShouldSendEmail   bool           `db:"should_send_email"`
 }
 
+type UserAccessTokens struct {
```
Comment (lhan-the-real) [inline internal/domain/users/domainService.go:36]: Nit: UserAccessToken (it's representing a single token)

File: internal/domain/users/usersStore_test.go [hunk 9]
```diff
@@ -214,6 +215,72 @@ var _ = Describe("UsersStore", func() {
 
 	})
 
+	Describe("User Access Tokens Related Methods", func() {
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore_test.go:218]: Nit:

```
	Describe("User Access Tokens", func() {
```

File: internal/domain/users/usersStore_test.go [hunk 10]
```diff
@@ -214,6 +215,72 @@ var _ = Describe("UsersStore", func() {
...
+				})
+			})
+
+			Context("invalid user access token and user agent are given", func() {
+				It("should successfully update the user access token", func() {
+					err := usersStore.InvalidateToken(context.Background(), "abc", "userAgent")
+					Expect(err).To(HaveOccurred())
+				})
+			})
+		})
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore_test.go:282]: Very fit for using `DescribeTable`

File: migrations/20220210055845_CreateUserAccessTokens.up.sql [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  id serial primary key,
+  user_id integer not null,
```
[the remark this answers is not in the document]
Comment (sleepyzzzzzz) [reply migrations/20220210055845_CreateUserAccessTokens.up.sql:3]: complete

File: migrations/20220210055845_CreateUserAccessTokens.up.sql (hunk 3 above)
Comment (sleepyzzzzzz) [reply migrations/20220210055845_CreateUserAccessTokens.up.sql:1]: complete

File: internal/domain/users/domainService.go (hunk 6 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/domainService.go:43]: yes

File: internal/domain/users/usersStore_test.go (hunk 7 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:240]: complete

File: internal/domain/users/domainService.go (hunk 8 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/domainService.go:36]: fix

File: internal/domain/users/usersStore_test.go (hunk 9 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:218]: fix

File: internal/domain/users/usersStore_test.go (hunk 10 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore_test.go:282]: put last two into DescribeTable

File: internal/domain/users/usersStore.go (hunk 5 above)
Comment (sleepyzzzzzz) [reply internal/domain/users/usersStore.go:340]: complete

File: internal/domain/utils/utils_test.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	"pdmgoauth/internal/domain/utils"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Utils", func() {
+	DescribeTable("HexMD5HashString", func(
```
Comment (lhan-the-real) [inline internal/domain/utils/utils_test.go:11]: love it.

File: internal/domain/users/usersStore.go [hunk 13]
```diff
@@ -302,3 +305,58 @@ func isValidEmailAddress(email string) bool {
...
+			return UserAccessToken{}, pdmerr.NewParent(err, "UsersStore: failed to create new user access token")
+		}
+
+		return UserAccessToken{}, pdmerr.NewParent(err, "UsersStore: failed to create new user access token")
+	}
+
+	return newUserToken, nil
+}
+
+// InvalidateToken sets the user's token to expire
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore.go:336]: Nit:
```
// InvalidateToken sets the user's token to expire now.
```

File: internal/domain/users/usersStore.go [hunk 14]
```diff
@@ -302,3 +305,58 @@ func isValidEmailAddress(email string) bool {
...
+	if err != nil {
+		return pdmerr.NewParent(
+			err,
+			"UsersStore: failed to update expire timestamp",
+		)
+	}
+
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return pdmerr.NewParent(err, "UsersStore: failed to log out user")
```
Comment (lhan-the-real) [inline internal/domain/users/usersStore.go:352]: If we are just invalidating the Token, let's use `failed to invalidate user access token`

**Review by lhan-the-real — APPROVED** [review]

**Pull request merged**